qmi/core/transport.py

```diff
@@ -686,22 +689,34 @@ def read_until(self, message_terminator: bytes, timeout: Optional[float]) -> byt
                 tremain = tstart + timeout - time.monotonic()
                 if tremain < 0:
                     nbuf = len(self._read_buffer)
-                    raise QMI_TimeoutException("Timeout after {} bytes without message terminator".format(nbuf))
+                    raise QMI_TimeoutException(f"Function timeout after {nbuf} bytes without message terminator")
                 self._safe_socket.settimeout(tremain)
+
             # Read from socket.
             try:
-                # NOTE: Reading up to 4096 bytes here.
+                # NOTE: Reading up to 512 bytes here for TCP, as larger packet size
+                #       causes issues with Tektronix 5014 AWG. For UDP read 4096.
                 #       Exact amount does not matter but a small power of 2 is recommended, e.g. 4096
                 #       by socket.recv() documentation.
-                b = self._safe_socket.recv(512)
-            except (BlockingIOError, socket.timeout):
+                if self._assert_addr:
+                    b, addr = self._safe_socket.recvfrom(4096)
+                else:
+                    b, addr = self._safe_socket.recvfrom(512)
+            except (BlockingIOError, socket.timeout, TimeoutError):
                 # timeout in socket.recv()
                 nbuf = len(self._read_buffer)
-                raise QMI_TimeoutException("Timeout after {} bytes without message terminator".format(nbuf))
+                raise QMI_TimeoutException(f"Socket timeout after {nbuf} bytes without message terminator")
+            except OSError:
```

Comment:
this exception does not seem to have a test

Reply:
Added a test